manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,729 @@
+package ipam
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/docker/libnetwork/ipamapi"
+	"github.com/docker/libnetwork/netlabel"
+
+	"github.com/docker/swarmkit/api"
+)
+
+// DrvRegistry is an interface defining
+type DrvRegistry interface {
+	IPAM(name string) (ipamapi.Ipam, *ipamapi.Capability)
+}
+
+// network is a private type used to hold the internal state of a network in
+// the IP Allocator
+type network struct {
+	// nw is a local cache of the network object found in the store
+	nw *api.Network
+
+	// pools is used to save the internal poolIDs needed when releasing a pool.
+	// It maps an ip address to a pool ID.
+	pools map[string]string
```

Comment:
One specific IP address. This is a design carried over from the previous allocator, which makes sense to me. Each IP address is allocated from a specific pool, and each pool has a pool id (returned by libnetwork as part of requesting the pool) which references that pool and is used for releasing. A network can have many pools, and the allocator will try to retrieve an address from each of them. If a pool is full, or you try to request a specific IP address that does not belong to that pool, then the allocator will try the next pool.

The end result of this is that there's no straightforward way to map an IP address to the pool it originated from. So, to circumvent this, you keep a mapping of every IP address to the pool it was allocated from, so that when you release it you can release to the correct pool.

But now that I get to thinking about it, I don't know what the performance implications of saving so many strings (one for each IP address) is. I think theoretically you SHOULD be able to compute which pool an IP address belongs to, and from there get the pool ID. I think that this way is more straightforward for now, though, and can be easily optimized later.

I'll add a comment explaining all this.